Allow explicit index file paths for inputs to CrosscheckFingerprints

rickymagner opened the pull request:
### Description

This PR addresses an issue regarding data storage solutions where indexed files can be separated from their indices. CrosscheckFingerprints will now allow for an `INPUT_INDEX_MAP` tsv file with two columns: the first being the list of input files, and the latter being the corresponding index file locations. One can also do the same for the `SECOND_INPUT` arguments using `SECOND_INPUT_INDEX_MAP`. 

Some tests have been added to test this functionality. This involved copying some test files into a separate dir from index files which were generated in a different directory. The test generates the `INPUT_INDEX_MAP` file on the fly so as to not rely on any absolute file paths. 

Although the tool still runs if an index is not provided, a new warning will be written if providing an index map without mapping each input to an index. If an index file path is typed incorrectly in the map, an error should be thrown by some dependency classes about not being able to import a valid index from the given path. 

----

### Checklist

#### Content
- [X] Added or modified tests to cover changes and any new functionality
- [X] Edited the README / documentation (if applicable)
- [x] All tests passing on Travis

#### Review
- [x] Final thumbs-up from reviewer
- [x] Rebase, squash and reword as applicable

rickymagner commented:
Here is the update to CrosscheckFingerprints as promised. I would appreciate a review from anyone who has time to take a look, especially any number of: @droazen, @yfarjoun, @meganshand, @kachulis, @lbergelson. Thanks!

meganshand commented on src/main/java/picard/fingerprint/FingerprintChecker.java:
```diff
@@ -679,6 +728,76 @@ public Map<FingerprintIdDetails, Fingerprint> fingerprintFiles(final Collection<
+                log.debug("Processed file: " + p.toUri().toString() + " (" + filesRead.get() + ")");
+
+                if (CheckFingerprint.fileContainsReads(p)) {
+                    if (indexPathMap.containsKey(p)) {
```
This might be cleaner to add to the method above rather than copying the surrounding code. You can check if `indexPathMap` is null to keep the "default index file" option.

rickymagner replied:
Refactored a bit to implement this

meganshand commented on src/test/java/picard/fingerprint/CrosscheckFingerprintsTest.java:
```diff
@@ -791,6 +792,99 @@ public void testCheckFiles(final List<File> files, final int expectedRetVal, fin
         doTest(args.toArray(new String[0]), metrics, expectedRetVal, numberOfSamples , CrosscheckMetric.DataType.FILE, ExpectAllMatch);
     }
 
+    // Test mostly same as last test, but with explicit index paths given
```
This test looks thorough for when the index files are in separate locations and provided, but you might also add a case where the index file map has missing elements to make sure the correct exception is thrown.

rickymagner replied:
Working on this still

rickymagner replied:
Just added a test that looks for the right exception in the case that `INPUT_INDEX_FORCE` is turned on but index paths are not provided (for files without indices next to them).

meganshand replied:
I like this solution. I think we should turn on INPUT_INDEX_FORCE in production since to safeguard against running without index files in the future.

rickymagner replied:
Definitely agree!

meganshand reviewed (commented):
Thanks @rickymagner!

rickymagner commented on src/main/java/picard/fingerprint/CrosscheckFingerprints.java:
```diff
@@ -479,8 +509,15 @@ protected int doWork() {
+        Map<FingerprintIdDetails, Fingerprint> fpMap = null;
+        if (INPUT_INDEX_MAP != null) {
+            final Map<FingerprintIdDetails, Fingerprint> uncappedFpMap = checker.fingerprintFiles(unrolledFiles, indexPathMap, NUM_THREADS, 1, TimeUnit.DAYS);
+            fpMap = capFingerprints(uncappedFpMap);
```
Some refactoring made this possible

rickymagner commented on src/main/java/picard/fingerprint/CrosscheckFingerprints.java:
```diff
@@ -713,6 +757,33 @@ private Map<String, String> getStringStringMap(final File sampleMapFile, final S
+                indexPath = IOUtil.getPath(entry.getValue());
+            }
+            catch(IOException e) {
+                throw new PicardException("Trouble reading file: " + entry.getValue(), e);
```
Added

rickymagner commented on src/main/java/picard/fingerprint/CrosscheckFingerprints.java:
```diff
@@ -253,6 +263,10 @@ public class CrosscheckFingerprints extends CommandLineProgram {
             doc = "One or more input files (or lists of files) with which to compare fingerprints.", minElements = 1)
     public List<String> INPUT;
 
+    @Argument(doc = "A tsv with two columns which maps the input files to corresponding indices; to be used when index " +
```
Changed

rickymagner commented on src/main/java/picard/fingerprint/CrosscheckFingerprints.java:
```diff
@@ -713,6 +757,33 @@ private Map<String, String> getStringStringMap(final File sampleMapFile, final S
+                inputPath = IOUtil.getPath(entry.getKey());
+            }
+            catch(IOException e) {
+                throw new PicardException("Trouble reading file: " + entry.getKey(), e);
```
Added

rickymagner commented:
Added another commit to allow for an `INPUT_FORCE_INDEX` argument. When toggled on, this checks that all files read for fingerprinting are using indices, either from those provided in `INPUT_INDEX_MAP` or found by the usual methods. The corresponding test for this PR has been updated to toggle this on (as it's off by default). The `SECOND_INPUT`s have the same option added.

I'm working on one more test to make sure when this option is on and inputs don't have a valid index, then the right exception should be thrown.

rickymagner commented:
In the meantime, I refactored the SAM file parsing a bit at [this line](https://github.com/broadinstitute/picard/pull/1823/files#diff-772be8f475d50e892ad98d9d2a755585cc05dac744ee866d3f7f7795d9dbb0ccR481). Previously this had a hot fix to address [this issue](https://github.com/broadinstitute/picard/issues/1175). 

It would be nice if someone with a bit more knowledge about `SamInputResource` could comment as to whether my code would fall into the same problem, or if it's avoided by using this different htsjdk code. Maybe one of @yfarjoun / @lbergelson  / @cmnbroad if they have a moment? Thanks

kachulis commented on src/main/java/picard/fingerprint/FingerprintChecker.java:
```diff
@@ -643,9 +673,26 @@ public Map<FingerprintIdDetails, Fingerprint> fingerprintFiles(final Collection<
 
                 if (CheckFingerprint.fileContainsReads(p)) {
-                    oneFileFingerprints = fingerprintSamFile(p, HaplotypeProbabilitiesFromSequence::new);
+                    if (indexPathMap != null && indexPathMap.containsKey(p)) {
```
You can pull the checking for the index outside of the if statement.  for example 
```java
final Path indexPath = indexPathMap == null ? null : indexPathMap.get(p);
                if (indexPathMap != null && indexPath == null) {
                    log.warn("Index map file provided, but no explicit index provided for " + p);
                } else if (indexPath != null) {
                    log.info("Using explicit index provided for " + p);
                }
```
and then you can use `indexPath` in the call to fingerprintSamFile or fingerprintVcf

rickymagner replied:
Good idea, thanks! I refactored a bit to clean up this block of code.

kachulis commented on src/main/java/picard/fingerprint/FingerprintChecker.java:
```diff
@@ -679,6 +726,7 @@ public Map<FingerprintIdDetails, Fingerprint> fingerprintFiles(final Collection<
         return retval;
     }
 
+
```
remove extra newline

rickymagner replied:
Done

rickymagner commented on src/main/java/picard/fingerprint/CrosscheckFingerprints.java:
```diff
@@ -173,31 +173,41 @@
-                        "A matrix output is also available to facilitate visual inspection of crosscheck results.\n " +
-                        "\n" +
+                        "A matrix output is also available to facilitate visual inspection of crosscheck results." +
+                        "<br /><br /> " +
```
This is because the [documentation](https://gatk.broadinstitute.org/hc/en-us/articles/5358894366363-CrosscheckFingerprints-Picard-) currently outputs the top as a big block of text. I looked at another tool documentation as a template and added the html line breaks to (hopefully) get better formatting when it gets pulled onto the web.

rickymagner commented on src/main/java/picard/fingerprint/CrosscheckFingerprints.java:
```diff
@@ -713,6 +757,33 @@ private Map<String, String> getStringStringMap(final File sampleMapFile, final S
+        final Map<String, String> indexStringMap = getStringStringMap(indexMapFile, inputFieldName);
+        final HashMap<Path, Path> indexPathMap = new HashMap<Path, Path>();
+        for (Map.Entry<String, String> entry: indexStringMap.entrySet()) {
+            Path inputPath = null;
```
This was a bit tricky because of the limited scope of the try block. It's possible to refactor and put the `try` block around the loop, but then you miss the rich exception messages about which file caused the exception. Unless you have a good idea of how to preserve that while getting `final` variables?

kachulis replied:
```suggestion
            final Path inputPath;
```

The point is that you can declare the variable (and make it final) without immediately assigning.  As long as you only assign it once, it can still be final.

rickymagner replied:
Oh neat, I didn't realize this was possible! Changed

rickymagner commented on src/main/java/picard/fingerprint/FingerprintChecker.java:
```diff
@@ -25,13 +25,7 @@
-import htsjdk.samtools.SamReader;
-import htsjdk.samtools.SamReaderFactory;
-import htsjdk.samtools.ValidationStringency;
+import htsjdk.samtools.*;
```
Huh, I think the IDE did that automatically somehow (or I pressed the wrong button somewhere). Let me revert that…

rickymagner commented on src/main/java/picard/fingerprint/FingerprintChecker.java:
```diff
@@ -445,6 +457,8 @@ public Map<FingerprintIdDetails, Fingerprint> fingerprintVcf(final Path vcfFile)
     }
 
+
+
```
Ah thanks!

rickymagner commented on src/main/java/picard/fingerprint/FingerprintChecker.java:
```diff
@@ -454,30 +468,40 @@ public Map<FingerprintIdDetails, Fingerprint> fingerprintVcf(final Path vcfFile)
+     */
+    public Map<FingerprintIdDetails, Fingerprint> fingerprintSamFile(final Path samFile, final Path indexPath, final boolean forceIndex,
+                                                                     final Function<HaplotypeBlock, HaplotypeProbabilities> blockToProbMapper) {
+        SamReader in = getSamReader(samFile, indexPath, forceIndex);
```
Changed

rickymagner commented on src/main/java/picard/fingerprint/FingerprintChecker.java:
```diff
@@ -454,30 +468,40 @@ public Map<FingerprintIdDetails, Fingerprint> fingerprintVcf(final Path vcfFile)
+     * Generates a Fingerprint per read group in the supplied SAM file using the loci provided in
+     * the interval list.
+     */
+    public Map<FingerprintIdDetails, Fingerprint> fingerprintSamFile(final Path samFile, final Path indexPath, final boolean forceIndex,
```
Oh right, good idea

rickymagner commented on src/main/java/picard/fingerprint/FingerprintChecker.java:
```diff
@@ -613,7 +637,8 @@ private FingerprintIdDetails createUnknownFP(final Path samFile, final SAMRecord
     public Map<String, Fingerprint> identifyContaminant(final Path samFile, final double contamination) {
 
-        final Map<FingerprintIdDetails, Fingerprint> fpIdDetailsMap = this.fingerprintSamFile(samFile, h -> new HaplotypeProbabilitiesFromContaminatorSequence(h, contamination));
+        final Map<FingerprintIdDetails, Fingerprint> fpIdDetailsMap = this.fingerprintSamFile(samFile, null,
```
Good call

rickymagner commented:
Added the last test as promised above for the `INPUT_INDEX_FORCE` option, and addressed the comments above. The only thing I'm waiting on is a bit more review + someone potentially addressing [this concern](https://github.com/broadinstitute/picard/pull/1823#issuecomment-1218145825).

meganshand commented on src/main/java/picard/fingerprint/CrosscheckFingerprints.java:
```diff
@@ -253,6 +263,13 @@ public class CrosscheckFingerprints extends CommandLineProgram {
+            "files are not located next to input files. First column must match the list of inputs. ", optional = true)
+    public File INPUT_INDEX_MAP;
+
+    @Argument(doc = "A boolean value to determine whether input files should only be parsed if index files are available.")
```
Maybe add a sentence here mentioning that it is potentially much slower to run this tool without index files? Basically giving some indication of why a user might choose to use this option.

rickymagner replied:
Good point. Added!

meganshand reviewed (approved):
This looks much cleaner and thank you for adding the additional tests!

kachulis commented on src/main/java/picard/fingerprint/FingerprintChecker.java:
```diff
@@ -176,6 +178,19 @@ public void setpLossofHet(final double pLossofHet) {
+
+        if (forceIndex && !reader.isQueryable()) {
+            throw new PicardException("Input VCF file " + vcfPath + " has no index while user required index to proceed.");
+        } else {
```
you don't need the else here.

rickymagner replied:
Good catch

rickymagner commented on src/main/java/picard/fingerprint/FingerprintChecker.java:
```diff
@@ -454,30 +474,46 @@ public Map<FingerprintIdDetails, Fingerprint> fingerprintVcf(final Path vcfFile)
+                .open(samResource);
+        if (forceIndex && !reader.isQueryable()) {
+            throw new PicardException("Input SAM file " + samFile + " has no index while user required index to proceed.");
+        } else {
```
Fixed

rickymagner commented on src/main/java/picard/fingerprint/CrosscheckFingerprints.java:
```diff
@@ -253,6 +263,15 @@ public class CrosscheckFingerprints extends CommandLineProgram {
+
+    @Argument(doc = "A boolean value to determine whether input files should only be parsed if index files are available. Without turning" +
+            "this option on, the tool will need to read through the entirety of input files without index files either provided via the INPUT_INDEX_MAP" +
+            " or locally accessible relative to the input, which significantly increases runtime.")
```
Clarified this

lbergelson commented on src/main/java/picard/fingerprint/FingerprintChecker.java:
```diff
@@ -454,30 +474,46 @@ public Map<FingerprintIdDetails, Fingerprint> fingerprintVcf(final Path vcfFile)
+    private SamReader getSamReader(final Path samFile, final Path indexPath, final boolean forceIndex) {
+        final SamInputResource samResource = SamInputResource.of(samFile);
+        if (indexPath != null) {
+            samResource.index(indexPath);
```
```suggestion
    samResource.index(indexPath, seekableChannelFunction);
```

rickymagner replied:
Thanks, just implemented this change and checked this should be equivalent to the old code. I left a comment pointing to the relevant issue in case someone tries to delete the `seekableChannelFunction` wrapper.

kachulis commented on src/main/java/picard/fingerprint/FingerprintChecker.java:
```diff
@@ -672,27 +684,19 @@ public Map<FingerprintIdDetails, Fingerprint> fingerprintFiles(final Collection<
                 log.debug("Processed file: " + p.toUri().toString() + " (" + filesRead.get() + ")");
 
+                // Determine whether valid index path given for this file
+                final Path indexPath = (indexPathMap != null && indexPathMap.containsKey(p)) ? indexPathMap.get(p) : null;
```
Don't need to check that p is in indexPathMap. ` indexPathMap.get(p)` will return null if it can't find p in the map

rickymagner replied:
Nice, removed the second boolean expression for the ternary check

kachulis commented on src/test/java/picard/fingerprint/CrosscheckFingerprintsTest.java:
```diff
@@ -886,6 +887,68 @@ public void testExplicitIndexPaths(final List<File> files, final List<File> indi
+        IOUtil.deleteOnExit(inputIndexMap);
+
+        // Zip files & indices together
+        List<List<String>> zipped_files = IntStream.range(0, Math.min(files.size(), indices.size()))
```
final

rickymagner replied:
Added

kachulis commented on src/test/java/picard/fingerprint/CrosscheckFingerprintsTest.java:
```diff
@@ -886,6 +887,68 @@ public void testExplicitIndexPaths(final List<File> files, final List<File> indi
+        }
+
+        // Set INPUT_INDEX_MAP to temp file created above
+        File INPUT_INDEX_MAP = new File(inputIndexMap.toString());
```
final

rickymagner replied:
Added

kachulis commented on src/main/java/picard/fingerprint/CrosscheckFingerprints.java:
```diff
@@ -253,6 +263,16 @@ public class CrosscheckFingerprints extends CommandLineProgram {
+            "files are not located next to input files. First column must match the list of inputs. ", optional = true)
+    public File INPUT_INDEX_MAP;
+
+    @Argument(doc = "A boolean value to determine whether input files should only be parsed if index files are available. Without turning" +
```
need a space after "turning"

rickymagner replied:
Fixed

kachulis commented on src/main/java/picard/fingerprint/CrosscheckFingerprints.java:
```diff
@@ -253,6 +263,16 @@ public class CrosscheckFingerprints extends CommandLineProgram {
+
+    @Argument(doc = "A boolean value to determine whether input files should only be parsed if index files are available. Without turning" +
+            "this option on, the tool will need to read through the entirety of input files without index files either provided via the INPUT_INDEX_MAP" +
+            " or locally accessible relative to the input, which significantly increases runtime. If set to true and no index is found for a file, an" +
```
space after final "an"

rickymagner replied:
Fixed

kachulis commented on src/main/java/picard/fingerprint/CrosscheckFingerprints.java:
```diff
@@ -280,6 +300,16 @@ public class CrosscheckFingerprints extends CommandLineProgram {
+            "files are not located next to second input files. First column must match the list of second inputs. ", optional = true)
+    public File SECOND_INPUT_INDEX_MAP;
+
+    @Argument(doc = "A boolean value to determine whether second input files should only be parsed if index files are available. Without turning" +
```
space at end of line

rickymagner replied:
Added

kachulis commented on src/main/java/picard/fingerprint/FingerprintChecker.java:
```diff
@@ -25,10 +25,12 @@
 package picard.fingerprint;
 
 import com.google.cloud.storage.contrib.nio.SeekableByteChannelPrefetcher;
+
```
remove newline

rickymagner replied:
Changed

kachulis commented on src/main/java/picard/fingerprint/FingerprintChecker.java:
```diff
@@ -185,8 +200,8 @@ public void setpLossofHet(final double pLossofHet) {
-    public Map<String, Fingerprint> loadFingerprints(final Path fingerprintFile, final String specificSample) {
-        final VCFFileReader reader = new VCFFileReader(fingerprintFile, false);
+    public Map<String, Fingerprint> loadFingerprints(final Path fingerprintFile, final Path indexPath, final boolean forceIndex, final String specificSample) {
+        final VCFFileReader reader = getVCFReader(fingerprintFile, indexPath, forceIndex);
```
what about 
```java
final VCFFileReader reader = new VCFFileReader(fingerprintFile, indexPath, forceIndex)
```

it looks like that constructor should handle a null indexPath correctly.

rickymagner replied:
This actually causes a bunch of tests to fail (apparently due to some null pointer exception). Although it seems like the method accepts `null` pointers, it looks like htsjdk handles that by assuming the index is next to the input file with extension `.tbi` which many of our test files don't have. I think it might be safer to leave it as is for now. This also allows for the custom PicardException to display when the user forces index file paths but they cannot be found.

kachulis commented on src/test/java/picard/fingerprint/CrosscheckFingerprintsTest.java:
```diff
@@ -791,6 +793,162 @@ public void testCheckFiles(final List<File> files, final int expectedRetVal, fin
+        IOUtil.deleteOnExit(inputIndexMap);
+
+        // Zip files & indices together
+        List<List<String>> zipped_files = IntStream.range(0, Math.min(files.size(), indices.size()))
```
final

and maybe refactor this code for writing index_index map file into its own method, since it's repeated a couple times.

rickymagner replied:
Done and done

kachulis reviewed (approved):
Just a couple of small things, other wise looks good.  Thanks @rickymagner!

rickymagner commented on src/test/java/picard/fingerprint/CrosscheckFingerprintsTest.java:
```diff
@@ -886,6 +887,68 @@ public void testExplicitIndexPaths(final List<File> files, final List<File> indi
+
+    @Test(dataProvider = "missingIndexForced", expectedExceptions = PicardException.class)
+    public void testMissingIndexForced(final List<File> files, final List<File> indices) throws IOException {
+        File metrics = File.createTempFile("Fingerprinting", "test.crosscheck_metrics");
```
Added

rickymagner commented on src/test/java/picard/fingerprint/CrosscheckFingerprintsTest.java:
```diff
@@ -886,6 +887,68 @@ public void testExplicitIndexPaths(final List<File> files, final List<File> indi
+
+        // Based on tabbedWrite method above
+        // Write input files & indices to tsv for input below
+        try (PrintWriter writer = new PrintWriter(Files.newBufferedWriter(inputIndexMap))) {
```
Added

kachulis commented on src/main/java/picard/fingerprint/FingerprintChecker.java:
```diff
@@ -454,30 +473,47 @@ public Map<FingerprintIdDetails, Fingerprint> fingerprintVcf(final Path vcfFile)
+        final SamInputResource samResource = SamInputResource.of(samFile);
+        if (indexPath != null) {
+            // Use of seekableChannelFunction here avoids issue: https://github.com/broadinstitute/picard/issues/1175
+            samResource.index(indexPath, seekableChannelFunction);
```
talked about this offline, but it looks like when using a SamInputResource, you need to explicitly give it an index if you want one to be used.  So you need to find the index if you haven't been given it explicitly.  See, as an example, the SamReader open method which takes a Path.  Would also be good to add a test to confirm that the index is used when it is available if not explicitly given as input.

rickymagner replied:
Good catch! Fixed and also added a test that should address this.

lbergelson reviewed (changes requested):
@rickymagner Sorry for the delay.  I have some comments.  The ones about the UI you can ignore if you feel like this is the cleanest way, but I think having a single input that can have a 3rd column with matched indexes like what David added recently in GATK (and we can maybe steal the loading code for) would be clearer.

I also am not sure I understand why you need 2 force index arguments, it seems extremely error prone.  I will 100% forget to force the second set of indexes every time.

lbergelson commented on src/test/java/picard/fingerprint/CrosscheckFingerprintsTest.java:
```diff
@@ -791,6 +793,149 @@ public void testCheckFiles(final List<File> files, final int expectedRetVal, fin
+        }
+
+        // Return temp file created above
+        return new File(inputIndexMap.toString());
```
```suggestion
        return inputIndexMap.toFile()
```

rickymagner replied:
Thanks!

rickymagner commented on src/main/java/picard/cmdline/CommandLineProgram.java:
```diff
@@ -71,7 +71,7 @@
  * To use:
  *
- * 1. Extend this class with a concrete class that is annotated with @COmmandLineProgramProperties, and has data members
+ * 1. Extend this class with a concrete class that is annotated with @CommandLineProgramProperties, and has data members
```
Hm, I made this change, which introduced an extra import if it's worth it.

lbergelson replied:
Yeah, that's fine I think.

rickymagner commented on src/main/java/picard/fingerprint/CrosscheckFingerprints.java:
```diff
@@ -173,31 +173,41 @@
-                        " - each samples from INPUT will only be compared to that same sample in SECOND_INPUT. \n" +
-                        " - MATRIX_OUTPUT is disabled. " +
-                        "\n" +
+                        "does the following:" +
```
It'd be much nicer to use markdown style text. Is that a more general organizational comment or something I can specifically change for this PR?

lbergelson replied:
No, this is just me whinging... there's almost certainly a path to change over to markdown javadoc gen, but it's not your job.

rickymagner commented on src/main/java/picard/fingerprint/CrosscheckFingerprints.java:
```diff
@@ -253,6 +263,16 @@ public class CrosscheckFingerprints extends CommandLineProgram {
     public List<String> INPUT;
 
+    @Argument(doc = "A tsv with two columns and no header which maps the input files to corresponding indices; to be used when index " +
+            "files are not located next to input files. First column must match the list of inputs. ", optional = true)
```
This strategy was originally thought through, but there were a few issues that came up that led me to abandon that path for this particular tool (after chatting with David R & Chris K about it too). Although it would be nice to match David's recent GATK PR paradigm, here are some problems with trying to implement it for this particular tool:

* There are actually three different ways to input sample information into CrosscheckFingerprints. There's the `INPUT` argument, the `INPUT_SAMPLE_MAP`, and the `INPUT_SAMPLE_FILE_MAP`. (And corresponding ones for the `SECOND_INPUT`…) These have different use cases for different parts of the downstream analysis.
* For this particular patch, we need to intervene at the moment when a VCF / SAM file is actually parsed by the tool to ensure index files can be provided. After inspecting the tool, I saw that only the `INPUT` arguments were ever fed to a parser. Hence, one can only intervene at the level of those file paths getting resolved. In fact, those are the only arguments which include paths at all.
* This meant it became necessary to allow the user to specify where the index files were for the `INPUT` paths specifically. The problem with trying to absorb this into the `INPUT_SAMPLE_MAP` or other similarly named file is that those don't actually contain a column with `INPUT` paths (hence why those get fed into `getStringStringMap` rather than a Path map). So we needed a totally new type of table which matched the file paths with index paths.

I agree longterm it'd be nice to adapt the UI a bit to agree with the other solution in the GATK tool, but it seems it would be a bit of a heavier lift than this was to implement.

lbergelson replied:
That's reasonable, thank you for explaining.  If we're going to be using these sample maps every where it would be nice of us to promote them to a more official thing and stick a sane parser in htsjdk or picard so we don't have to reinvent them repeatedly, but not the job of this pr.

rickymagner commented on src/main/java/picard/fingerprint/CrosscheckFingerprints.java:
```diff
@@ -253,6 +263,16 @@ public class CrosscheckFingerprints extends CommandLineProgram {
+            "this option on, the tool will need to read through the entirety of input files without index files either provided via the INPUT_INDEX_MAP" +
+            " or locally accessible relative to the input, which significantly increases runtime. If set to true and no index is found for a file, an " +
+            "exception will be thrown.")
+    public Boolean INPUT_FORCE_INDEX = false;
```
As much as I prefer "indices" over "indexes" in this context, how about we avoid it altogether with the arguably more clear `REQUIRE_INDEX_FILES`?

lbergelson replied:
Good solution!

rickymagner commented on src/main/java/picard/fingerprint/CrosscheckFingerprints.java:
```diff
@@ -280,6 +300,16 @@ public class CrosscheckFingerprints extends CommandLineProgram {
+            "files are not located next to second input files. First column must match the list of second inputs. ", optional = true)
+    public File SECOND_INPUT_INDEX_MAP;
+
+    @Argument(doc = "A boolean value to determine whether second input files should only be parsed if index files are available. Without turning " +
```
Probably not. I changed it so there's just one flag to require index files used in both cases.

rickymagner commented on src/main/java/picard/fingerprint/CrosscheckFingerprints.java:
```diff
@@ -253,6 +263,16 @@ public class CrosscheckFingerprints extends CommandLineProgram {
             doc = "One or more input files (or lists of files) with which to compare fingerprints.", minElements = 1)
```
I agree it's messy, but see my above comment on this choice.

rickymagner commented on src/main/java/picard/fingerprint/CrosscheckFingerprints.java:
```diff
@@ -253,6 +263,16 @@ public class CrosscheckFingerprints extends CommandLineProgram {
+            "this option on, the tool will need to read through the entirety of input files without index files either provided via the INPUT_INDEX_MAP" +
+            " or locally accessible relative to the input, which significantly increases runtime. If set to true and no index is found for a file, an " +
+            "exception will be thrown.")
+    public Boolean INPUT_FORCE_INDEX = false;
```
Oh, I assumed if it didn't give an error that was the only way to do it. Changed to the nicer lowercase.

rickymagner commented on src/main/java/picard/fingerprint/CrosscheckFingerprints.java:
```diff
@@ -253,6 +263,16 @@ public class CrosscheckFingerprints extends CommandLineProgram {
 
+    @Argument(doc = "A tsv with two columns and no header which maps the input files to corresponding indices; to be used when index " +
+            "files are not located next to input files. First column must match the list of inputs. ", optional = true)
+    public File INPUT_INDEX_MAP;
```
It looks like changing these away from `File` causes a number of other problems, especially `PicardHtsPath`. But even `Path` causes a few downstream changes that would probably be fixed by writing `mypath.toFile()` in a bunch of places. If you think this is a worthwhile pursuit, I can try to hunt down the appropriate changes for this. But it might be best saved for a full rewrite where all of the inputs are appropriately updated.

lbergelson replied:
If it's a pain we can put it off.  Thanks for taking a look.

rickymagner commented on src/main/java/picard/fingerprint/CrosscheckFingerprints.java:
```diff
@@ -713,6 +763,33 @@ private Map<String, String> getStringStringMap(final File sampleMapFile, final S
+    private Map<Path, Path> getPathPathMap(final File indexMapFile, final String inputFieldName) {
+        final Map<String, String> indexStringMap = getStringStringMap(indexMapFile, inputFieldName);
+        final HashMap<Path, Path> indexPathMap = new HashMap<Path, Path>();
+        for (Map.Entry<String, String> entry: indexStringMap.entrySet()) {
```
Thanks for this advice. Changed to LinkedHashMap in both methods.

lbergelson replied:
I will continue my quest to abolish java HashMap until it or I am defeated!

rickymagner commented on src/main/java/picard/fingerprint/CrosscheckFingerprints.java:
```diff
@@ -713,6 +763,33 @@ private Map<String, String> getStringStringMap(final File sampleMapFile, final S
+            try {
+                inputPath = IOUtil.getPath(entry.getKey());
+            }
+            catch(IOException e) {
```
Nice, that looks better!

rickymagner commented on src/main/java/picard/fingerprint/CrosscheckFingerprints.java:
```diff
@@ -713,6 +763,33 @@ private Map<String, String> getStringStringMap(final File sampleMapFile, final S
+            final Path inputPath;
+            final Path indexPath;
+
+            // Attempt to read input path
```
Hmm, how about "process index path" instead of "read index path"? Made this change

rickymagner commented on src/main/java/picard/fingerprint/CrosscheckFingerprints.java:
```diff
@@ -713,6 +763,33 @@ private Map<String, String> getStringStringMap(final File sampleMapFile, final S
+                inputPath = IOUtil.getPath(entry.getKey());
+            }
+            catch(IOException e) {
+                throw new PicardException("Trouble reading file: " + entry.getKey() + " for field " + inputFieldName, e);
```
Hm, I think `field` here refers to the argument field, e.g. `INPUT_INDEX_MAP` or `SECOND_INPUT_INDEX_MAP`. Maybe that should change to `argument` instead?

rickymagner replied:
Just changed this for better readability, thanks!

rickymagner commented on src/main/java/picard/fingerprint/FingerprintChecker.java:
```diff
@@ -176,6 +179,20 @@ public void setpLossofHet(final double pLossofHet) {
+     * Loads VCF reader from path, and checks if index available when forced.
+     */
+    @VisibleForTesting
+    VCFFileReader getVCFReader(final Path vcfPath, final Path indexPath, final boolean forceIndex) {
```
Made this one static, but for the other VisibleForTesting SAMReader method, it requires `referenceFasta` which is set by the constructor, so I don't think that can be made static.

rickymagner commented on src/main/java/picard/fingerprint/FingerprintChecker.java:
```diff
@@ -430,10 +447,15 @@ public IntervalList getLociToGenotype(final Collection<Fingerprint> fingerprints
     }
 
+    // For backwards compatibility with previous interface
     public Map<FingerprintIdDetails, Fingerprint> fingerprintVcf(final Path vcfFile) {
```
Added

rickymagner commented on src/test/java/picard/fingerprint/CrosscheckFingerprintsTest.java:
```diff
@@ -791,6 +793,149 @@ public void testCheckFiles(final List<File> files, final int expectedRetVal, fin
+        final Path inputIndexMap = File.createTempFile("input_index_map", ".tsv").toPath();
+        IOUtil.deleteOnExit(inputIndexMap);
+
+        // Zip files & indices together
```
Much cleaner, thanks!

lbergelson commented:
@rickymagner Thanks for addressing those.  Some comments are more aspirational than practical.  Looks good to me.

lbergelson reviewed (approved):
👍

kachulis reviewed (approved):
Looks good, thanks @rickymagner!
